File: contracts/scripts/validate-schemas.js
#!/usr/bin/env node
/**
 * CI script to validate all JSON schemas compile correctly with ajv.
 */

const Ajv2020 = require('ajv/dist/2020');
const addFormats = require('ajv-formats');
const fs = require('fs');
const path = require('path');

const SCHEMA_DIR = path.join(__dirname, '..', 'schemas', 'v1');

// Schema files to validate
const SCHEMAS = [
  'document-ref.schema.json',
  'document-ir.schema.json',
  'chunk-record.schema.json',
  'retrieval-plan.schema.json',
  'trace.schema.json',
  'golden-record.schema.json',
  'embedding.schema.json',
];

function loadSchema(filename) {
  const filepath = path.join(SCHEMA_DIR, filename);
  const content = fs.readFileSync(filepath, 'utf-8');
  return JSON.parse(content);
}

function validateSchema(ajv, filename) {
  console.log(`Validating: ${filename}`);
  
  try {
    const schema = loadSchema(filename);
    
    // For embedding.schema.json, validate nested schemas
    if (filename === 'embedding.schema.json') {
      if (schema.embedding_templates_spec?.schema) {
        const compiled = ajv.compile(schema.embedding_templates_spec.schema);
        console.log(`  ✓ embedding_templates_spec schema compiles`);
      }
      if (schema.multi_view_vector_record_spec?.schema) {
        const compiled = ajv.compile(schema.multi_view_vector_record_spec.schema);
        console.log(`  ✓ multi_view_vector_record_spec schema compiles`);
      }
      return true;
    }
    
    // Compile schema (validates it's well-formed)
    const compiled = ajv.compile(schema);
    console.log(`  ✓ Schema compiles successfully`);
    return true;
  } catch (error) {
    console.log(`  ✗ FAILED: ${error.message}`);
    return false;
  }
}

function main() {
  console.log('=' .repeat(60));
  console.log('  NPR Contracts - JSON Schema Validation (ajv)');
  console.log('=' .repeat(60));
  console.log();

  // Create ajv instance with JSON Schema draft-2020-12
  const ajv = new Ajv2020({
    strict: false,
    allErrors: true,
    verbose: true,
  });
  addFormats(ajv);

  const results = [];

  for (const schema of SCHEMAS) {
    const schemaPath = path.join(SCHEMA_DIR, schema);
    if (fs.existsSync(schemaPath)) {
      results.push(validateSchema(ajv, schema));
    } else {
      console.log(`Skipping (not found): ${schema}`);
      results.push(true); // Don't fail for missing files
    }
  }

  console.log();
  console.log('=' .repeat(60));
  console.log('  SUMMARY');
  console.log('=' .repeat(60));

  const passed = results.filter(r => r).length;
  const total = results.length;

  console.log(`  Results: ${passed}/${total} schemas validated`);
  console.log('=' .repeat(60));

  process.exit(results.every(r => r) ? 0 : 1);
}

main();

File: contracts/scripts/validate_schemas.py
#!/usr/bin/env python3
"""CI script to validate golden sets and embedding records against schemas.

Usage:
    python validate_schemas.py --golden path/to/golden.json
    python validate_schemas.py --embeddings path/to/embeddings.json
    python validate_schemas.py --all  # validate all known files
"""

import argparse
import json
import sys
from pathlib import Path

# Add parent to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent / "python"))

from npr_contracts import (
    validate_golden_set,
    validate_golden_record,
    validate_embedding_templates,
    validate_vector_record,
    SchemaValidationError,
)


def validate_golden_file(filepath: Path) -> bool:
    """Validate a golden set JSON file."""
    print(f"Validating golden set: {filepath}")
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            data = json.load(f)
        
        if "records" in data:
            validate_golden_set(data)
            print(f"  ✓ Valid ({len(data['records'])} records)")
        else:
            validate_golden_record(data)
            print(f"  ✓ Valid (single record)")
        return True
    except SchemaValidationError as e:
        print(f"  ✗ FAILED: {e}")
        return False
    except Exception as e:
        print(f"  ✗ ERROR: {e}")
        return False


def validate_embedding_file(filepath: Path) -> bool:
    """Validate an embedding templates or vector records file."""
    print(f"Validating embedding file: {filepath}")
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            data = json.load(f)
        
        if "embedding_templates_spec" in data:
            templates = data["embedding_templates_spec"].get("defaults", {})
            validate_embedding_templates(templates)
            print(f"  ✓ Valid embedding templates")
        elif "views" in data:
            validate_vector_record(data)
            print(f"  ✓ Valid vector record")
        else:
            print(f"  ? Unknown embedding file format")
            return False
        return True
    except SchemaValidationError as e:
        print(f"  ✗ FAILED: {e}")
        return False
    except Exception as e:
        print(f"  ✗ ERROR: {e}")
        return False


def main():
    parser = argparse.ArgumentParser(description="Validate NPR schema files")
    parser.add_argument("--golden", type=Path, nargs="*", help="Golden set files to validate")
    parser.add_argument("--embeddings", type=Path, nargs="*", help="Embedding files to validate")
    parser.add_argument("--all", action="store_true", help="Validate all known schema files")
    args = parser.parse_args()

    results = []
    
    if args.golden:
        for f in args.golden:
            results.append(validate_golden_file(f))
    
    if args.embeddings:
        for f in args.embeddings:
            results.append(validate_embedding_file(f))
    
    if args.all:
        # Look for files in contracts/schemas/v1/
        schema_dir = Path(__file__).parent.parent / "schemas" / "v1"
        golden_example = schema_dir / "golden-example.json"
        if golden_example.exists():
            results.append(validate_golden_file(golden_example))
        
        embedding_file = schema_dir / "embedding.schema.json"
        if embedding_file.exists():
            results.append(validate_embedding_file(embedding_file))

    if not results:
        parser.print_help()
        sys.exit(1)
    
    # Summary
    print()
    passed = sum(results)
    total = len(results)
    print(f"Results: {passed}/{total} passed")
    
    sys.exit(0 if all(results) else 1)


if __name__ == "__main__":
    main()
